// kiotviet-client-sdk/dist/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KiotVietClient = void 0;
const axios_1 = __importDefault(require("axios"));
const constants_1 = require("./config/constants");
const token_manager_1 = require("./services/token-manager");
const interceptor_service_1 = require("./services/interceptor-service");
const categories_1 = require("./resources/categories");
const customers_1 = require("./resources/customers");
const products_1 = require("./resources/products");
const orders_1 = require("./resources/orders");
const invoices_1 = require("./resources/invoices");
const webhooks_1 = require("./resources/webhooks");
const users_1 = require("./resources/users");
const purchase_orders_1 = require("./resources/purchase-orders");
const branches_1 = require("./resources/branches");
const bank_accounts_1 = require("./resources/bank-accounts");
const price_books_1 = require("./resources/price-books");
const suppliers_1 = require("./resources/suppliers");
const transfers_1 = require("./resources/transfers");
const surcharges_1 = require("./resources/surcharges");
const cash_flow_1 = require("./resources/cash-flow");
const returns_1 = require("./resources/returns");
const vouchers_1 = require("./resources/vouchers");
const sales_channels_1 = require("./resources/sales-channels");
const trademarks_1 = require("./resources/trademarks");
const settings_1 = require("./resources/settings");
class KiotVietClient {
    constructor(config) {
        this.validateConfig(config);
        this.config = this.initializeConfig(config);
        this.apiClient = this.createApiClient();
        this.tokenManager = new token_manager_1.TokenManager(this.config);
        // Initialize interceptors
        new interceptor_service_1.InterceptorService(this.apiClient, this.tokenManager, this.config.retailerName);
        // Initialize resource handlers
        this.customers = new customers_1.CustomerHandler(this);
        this.categories = new categories_1.CategoryHandler(this);
        this.orders = new orders_1.OrderHandler(this);
        this.products = new products_1.ProductHandler(this);
        this.invoices = new invoices_1.InvoiceHandler(this);
        this.webhooks = new webhooks_1.WebhookHandler(this);
        this.users = new users_1.UserHandler(this);
        this.purchaseOrders = new purchase_orders_1.PurchaseOrderHandler(this);
        this.branches = new branches_1.BranchHandler(this);
        this.bankAccounts = new bank_accounts_1.BankAccountHandler(this);
        this.priceBooks = new price_books_1.PriceBookHandler(this);
        this.suppliers = new suppliers_1.SupplierHandler(this);
        this.transfers = new transfers_1.TransferHandler(this);
        this.surcharges = new surcharges_1.SurchargeHandler(this);
        this.cashFlow = new cash_flow_1.CashFlowHandler(this);
        this.returns = new returns_1.ReturnsHandler(this);
        this.vouchers = new vouchers_1.VouchersHandler(this);
        this.salesChannels = new sales_channels_1.SalesChannelsHandler(this);
        this.trademarks = new trademarks_1.TrademarksHandler(this);
        this.settings = new settings_1.SettingsHandler(this);
    }
    validateConfig(config) {
        if (!config.clientId || !config.clientSecret || !config.retailerName) {
            throw new Error('clientId, clientSecret, and retailerName are required');
        }
    }
    initializeConfig(config) {
        var _a;
        return Object.assign(Object.assign({}, config), { baseUrl: config.baseUrl || constants_1.API_CONSTANTS.DEFAULT_BASE_URL, tokenUrl: config.tokenUrl || constants_1.API_CONSTANTS.DEFAULT_TOKEN_URL, apiVersion: config.apiVersion || constants_1.API_CONSTANTS.DEFAULT_API_VERSION, timeout: (_a = config.timeout) !== null && _a !== void 0 ? _a : constants_1.API_CONSTANTS.DEFAULT_TIMEOUT });
    }
    createApiClient() {
        return axios_1.default.create({
            baseURL: this.config.baseUrl.replace(/\/$/, ''),
            timeout: this.config.timeout,
        });
    }
    /**
     * Manually trigger a token refresh
     * @returns Promise<string> The new access token
     */
    refreshToken() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.tokenManager.refreshToken();
        });
    }
    /**
     * Make a GET request
     * @param url The URL to make the request to
     * @param config Optional axios request configuration
     * @returns Promise with the response data
     */
    get(url, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.apiClient.get(url, config);
            return response.data;
        });
    }
    /**
     * Make a POST request
     * @param url The URL to make the request to
     * @param data The data to send in the request body
     * @param config Optional axios request configuration
     * @returns Promise with the response data
     */
    post(url, data, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.apiClient.post(url, data, config);
            return response.data;
        });
    }
    /**
     * Make a DELETE request
     * @param url The URL to make the request to
     * @param config Optional axios request configuration
     * @returns Promise with the response data
     */
    delete(url, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.apiClient.delete(url, config);
            return response.data;
        });
    }
    /**
     * Make a PUT request
     * @param url The URL to make the request to
     * @param data The data to send in the request body
     * @param config Optional axios request configuration
     * @returns Promise with the response data
     */
    put(url, data, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.apiClient.put(url, data, config);
            return response.data;
        });
    }
}
exports.KiotVietClient = KiotVietClient;
