// kiotviet-client-sdk/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = exports.ForbiddenError = exports.RateLimitError = exports.NotFoundError = exports.AuthenticationError = exports.KiotVietApiError = void 0;
class KiotVietApiError extends Error {
    constructor(message, statusCode, errorStatus, responseBody) {
        super(message);
        this.name = 'KiotVietApiError';
        this.statusCode = statusCode;
        this.errorCode = errorStatus === null || errorStatus === void 0 ? void 0 : errorStatus.errorCode;
        // Use KiotViet message if available, otherwise the generic message
        this.errorMessage = (errorStatus === null || errorStatus === void 0 ? void 0 : errorStatus.message) || message;
        this.responseBody = responseBody;
        // Maintains proper stack trace (only available on V8)
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, KiotVietApiError);
        }
    }
}
exports.KiotVietApiError = KiotVietApiError;
class AuthenticationError extends KiotVietApiError {
    constructor(message, responseBody) {
        super(`Authentication Failed: ${message}`, 401, undefined, responseBody);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
class NotFoundError extends KiotVietApiError {
    constructor(message, responseBody) {
        super(`Resource Not Found: ${message}`, 404, undefined, responseBody);
        this.name = 'NotFoundError';
    }
}
exports.NotFoundError = NotFoundError;
class RateLimitError extends KiotVietApiError {
    constructor(message, responseBody) {
        super(`Rate Limit Exceeded: ${message}`, 429, undefined, responseBody);
        this.name = 'RateLimitError';
    }
}
exports.RateLimitError = RateLimitError;
class ForbiddenError extends KiotVietApiError {
    constructor(message, responseBody) {
        super(`Access Forbidden: ${message}`, 403, undefined, responseBody);
        this.name = 'ForbiddenError';
    }
}
exports.ForbiddenError = ForbiddenError;
class ValidationError extends KiotVietApiError {
    constructor(message, errorStatus, responseBody) {
        super(message, 400, errorStatus, responseBody);
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
