// kiotviet-client-sdk/dist/services/interceptor-service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterceptorService = void 0;
const error_handler_1 = require("./error-handler");
const constants_1 = require("../config/constants");
class InterceptorService {
    constructor(apiClient, tokenManager, retailerName) {
        this.apiClient = apiClient;
        this.tokenManager = tokenManager;
        this.retailerName = retailerName;
        this.setupInterceptors();
    }
    setupInterceptors() {
        this.setupRequestInterceptor();
        this.setupResponseInterceptor();
    }
    setupRequestInterceptor() {
        this.apiClient.interceptors.request.use((request) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const token = yield this.tokenManager.getValidToken();
            request.headers['Authorization'] = `Bearer ${token}`;
            request.headers['Retailer'] = this.retailerName;
            if (Object.values(constants_1.HTTP_METHODS).includes((_a = request.method) === null || _a === void 0 ? void 0 : _a.toUpperCase()) &&
                request.data &&
                !request.headers['Content-Type']) {
                request.headers['Content-Type'] = constants_1.CONTENT_TYPES.JSON;
            }
            return request;
        }), (error) => Promise.reject(error));
    }
    setupResponseInterceptor() {
        this.apiClient.interceptors.response.use((response) => response, (error) => __awaiter(this, void 0, void 0, function* () {
            return Promise.reject(error_handler_1.ErrorHandler.handleRequestError(error));
        }));
    }
}
exports.InterceptorService = InterceptorService;
