// kiotviet-client-sdk/dist/types/webhook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookEvent = void 0;
/**
 * Enum định nghĩa tất cả các loại sự kiện webhook hỗ trợ bởi KiotViet
 */
var WebhookEvent;
(function (WebhookEvent) {
    // Hàng hóa (Product)
    WebhookEvent["ProductCreated"] = "product.created";
    WebhookEvent["ProductUpdated"] = "product.updated";
    WebhookEvent["ProductDeleted"] = "product.deleted";
    // Danh mục (Category)
    WebhookEvent["CategoryCreated"] = "category.created";
    WebhookEvent["CategoryUpdated"] = "category.updated";
    WebhookEvent["CategoryDeleted"] = "category.deleted";
    // Khách hàng (Customer)
    WebhookEvent["CustomerCreated"] = "customer.created";
    WebhookEvent["CustomerUpdated"] = "customer.updated";
    WebhookEvent["CustomerDeleted"] = "customer.deleted";
    // Đơn hàng (Order)
    WebhookEvent["OrderCreated"] = "order.created";
    WebhookEvent["OrderUpdated"] = "order.updated";
    WebhookEvent["OrderDeleted"] = "order.deleted";
    // Hóa đơn (Invoice)
    WebhookEvent["InvoiceCreated"] = "invoice.created";
    WebhookEvent["InvoiceUpdated"] = "invoice.updated";
    WebhookEvent["InvoiceDeleted"] = "invoice.deleted";
    // Tồn kho (Stock)
    WebhookEvent["StockUpdated"] = "stock.update";
    // Bảng giá (PriceBook)
    WebhookEvent["PriceBookUpdated"] = "pricebook.update";
    WebhookEvent["PriceBookDeleted"] = "pricebook.delete";
    WebhookEvent["PriceBookDetailUpdated"] = "pricebookdetail.update";
    WebhookEvent["PriceBookDetailDeleted"] = "pricebookdetail.delete";
    // Chi nhánh (Branch)
    WebhookEvent["BranchUpdated"] = "branch.update";
    WebhookEvent["BranchDeleted"] = "branch.delete";
})(WebhookEvent || (exports.WebhookEvent = WebhookEvent = {}));
