// kiotviet-client-sdk/dist/resources/branches.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BranchHandler = void 0;
class BranchHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List branches with optional filtering and pagination
     * @param params Filter and pagination parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/branches', { params });
            return response.data;
        });
    }
    /**
     * Get a branch by its ID
     * @param branchId The ID of the branch to retrieve
     */
    getById(branchId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/branches/${branchId}`);
            return response.data;
        });
    }
    /**
     * Create a new branch
     * @param branchData The branch data to create
     */
    create(branchData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/branches', branchData);
            return response.data;
        });
    }
    /**
     * Update an existing branch
     * @param branchId The ID of the branch to update
     * @param branchData The branch data to update
     */
    update(branchId, branchData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/branches/${branchId}`, branchData);
            return response.data;
        });
    }
    /**
     * Delete a branch
     * @param branchId The ID of the branch to delete
     */
    delete(branchId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.delete(`/branches/${branchId}`);
        });
    }
}
exports.BranchHandler = BranchHandler;
