// kiotviet-client-sdk/dist/resources/categories.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CategoryHandler = void 0;
class CategoryHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List categories with optional filtering and pagination
     * @param params Filter and pagination parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/categories', { params });
            return response.data;
        });
    }
    /**
     * Get a category by its ID
     * @param categoryId The ID of the category to retrieve
     */
    getById(categoryId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/categories/${categoryId}`);
            return response.data;
        });
    }
    /**
     * Create a new category
     * @param categoryData The category data to create
     */
    create(categoryData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/categories', categoryData);
            return response.data;
        });
    }
    /**
     * Update an existing category
     * @param categoryId The ID of the category to update
     * @param categoryData The category data to update
     */
    update(categoryId, categoryData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/categories/${categoryId}`, categoryData);
            return response.data;
        });
    }
    /**
     * Delete a category
     * @param categoryId The ID of the category to delete
     */
    delete(categoryId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.delete(`/categories/${categoryId}`);
        });
    }
    /**
     * Get categories with hierarchical structure
     * @param params Filter and pagination parameters
     */
    getHierarchical() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/categories', {
                params: Object.assign(Object.assign({}, params), { hierarchicalData: true }),
            });
            return response.data;
        });
    }
}
exports.CategoryHandler = CategoryHandler;
