// kiotviet-client-sdk/dist/resources/customers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomerHandler = void 0;
const errors_1 = require("../errors");
class CustomerHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List customers with optional filtering
     * @param params Filter parameters (pageSize, currentItem)
     * Documentation: GET /customers
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/customers', { params });
            return response.data;
        });
    }
    /**
     * Get a customer by their ID
     * @param customerId The ID of the customer to retrieve
     * Documentation: GET /customers/{id}
     */
    getById(customerId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/customers/${customerId}`);
            return response.data;
        });
    }
    /**
     * Create a new customer
     * @param customerData The customer data to create
     * Documentation: POST /customers
     */
    create(customerData) {
        return __awaiter(this, void 0, void 0, function* () {
            // Validate required fields
            if (!customerData.name) {
                throw new errors_1.ValidationError('Customer name is required');
            }
            const response = yield this.client.apiClient.post('/customers', customerData);
            return response.data;
        });
    }
    /**
     * Search customers by keyword
     * @param query Search query
     * @param params Additional filter parameters
     */
    search(query_1) {
        return __awaiter(this, arguments, void 0, function* (query, params = {}) {
            const response = yield this.client.apiClient.get('/customers', {
                params: Object.assign(Object.assign({}, params), { keyword: query }),
            });
            return response.data;
        });
    }
    /**
     * Get customers by group ID
     * @param groupId The ID of the customer group
     * @param params Additional filter parameters
     */
    getByGroup(groupId_1) {
        return __awaiter(this, arguments, void 0, function* (groupId, params = {}) {
            const response = yield this.client.apiClient.get('/customers', {
                params: Object.assign(Object.assign({}, params), { customerGroupId: groupId }),
            });
            return response.data;
        });
    }
    /**
     * Get customer by contact number
     * @param contactNumber The customer's contact number
     */
    getByContactNumber(contactNumber) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get('/customers', {
                params: {
                    contactNumber,
                    pageSize: 1,
                },
            });
            return response.data.data.length > 0 ? response.data.data[0] : null;
        });
    }
}
exports.CustomerHandler = CustomerHandler;
