// kiotviet-client-sdk/dist/resources/invoices.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoiceHandler = void 0;
const types_1 = require("../types");
class InvoiceHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List invoices with optional filtering
     * @param params Filter parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/invoices', { params });
            return response.data;
        });
    }
    /**
     * Get an invoice by its ID
     * @param invoiceId The ID of the invoice to retrieve
     */
    getById(invoiceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/invoices/${invoiceId}`);
            return response.data;
        });
    }
    /**
     * Create a new invoice
     * @param invoiceData The invoice data to create
     */
    create(invoiceData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/invoices', invoiceData);
            return response.data;
        });
    }
    /**
     * Update an existing invoice
     * @param invoiceId The ID of the invoice to update
     * @param invoiceData The invoice data to update
     */
    update(invoiceId, invoiceData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/invoices/${invoiceId}`, Object.assign({ id: invoiceId }, invoiceData));
            return response.data;
        });
    }
    /**
     * Cancel an invoice
     * @param invoiceId The ID of the invoice to cancel
     * @param reason Optional cancellation reason
     */
    cancel(invoiceId, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/invoices/${invoiceId}`, {
                id: invoiceId,
                status: types_1.InvoiceStatus.Cancelled,
                description: reason,
            });
            return response.data;
        });
    }
    /**
     * Get invoices by date range
     * @param fromDate Start date (YYYY-MM-DD)
     * @param toDate End date (YYYY-MM-DD)
     * @param params Additional filter parameters
     */
    getByDateRange(fromDate_1, toDate_1) {
        return __awaiter(this, arguments, void 0, function* (fromDate, toDate, params = {}) {
            const response = yield this.client.apiClient.get('/invoices', {
                params: Object.assign(Object.assign({}, params), { fromPurchaseDate: fromDate, toPurchaseDate: toDate }),
            });
            return response.data;
        });
    }
    /**
     * Get invoices by customer
     * @param customerIdentifier Customer's phone number or code
     * @param params Additional filter parameters
     */
    getByCustomer(customerIdentifier_1) {
        return __awaiter(this, arguments, void 0, function* (customerIdentifier, params = {}) {
            // Try to determine if the identifier is a phone number
            const isPhone = /^\d+$/.test(customerIdentifier);
            const response = yield this.client.apiClient.get('/invoices', {
                params: Object.assign(Object.assign({}, params), (isPhone ? { customerPhone: customerIdentifier } : { customerCode: customerIdentifier })),
            });
            return response.data;
        });
    }
    /**
     * Get invoices by order
     * @param orderId The ID of the order
     * @param params Additional filter parameters
     */
    getByOrder(orderId_1) {
        return __awaiter(this, arguments, void 0, function* (orderId, params = {}) {
            const response = yield this.client.apiClient.get('/invoices', {
                params: Object.assign(Object.assign({}, params), { orderId }),
            });
            return response.data;
        });
    }
    /**
     * Get an invoice by its code
     * @param code The code of the invoice to retrieve
     */
    getByCode(code) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/invoices/code/${code}`);
            return response.data;
        });
    }
    /**
     * Delete/void an invoice
     * @param invoiceId The ID of the invoice to delete
     * @param isVoidPayment Whether to void the associated payment
     */
    delete(invoiceId_1) {
        return __awaiter(this, arguments, void 0, function* (invoiceId, isVoidPayment = false) {
            yield this.client.apiClient.delete('/invoices', {
                params: {
                    id: invoiceId,
                    isVoidPayment: isVoidPayment,
                },
            });
        });
    }
}
exports.InvoiceHandler = InvoiceHandler;
