// kiotviet-client-sdk/dist/resources/orders.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderHandler = void 0;
const order_1 = require("../types/order");
class OrderHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List orders with optional filtering
     * @param params Filter parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/orders', { params });
            return response.data;
        });
    }
    /**
     * Get an order by its ID
     * @param orderId The ID of the order to retrieve
     */
    getById(orderId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/orders/${orderId}`);
            return response.data;
        });
    }
    /**
     * Create a new order
     * @param orderData The order data to create
     */
    create(orderData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/orders', orderData);
            return response.data;
        });
    }
    /**
     * Update an existing order
     * @param orderId The ID of the order to update
     * @param orderData The order data to update
     */
    update(orderId, orderData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/orders/${orderId}`, Object.assign({ id: orderId }, orderData));
            return response.data;
        });
    }
    /**
     * Cancel an order
     * @param orderId The ID of the order to cancel
     * @param reason Optional cancellation reason
     */
    cancel(orderId, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/orders/${orderId}`, {
                id: orderId,
                status: order_1.OrderStatus.Cancelled,
                description: reason,
            });
            return response.data;
        });
    }
    /**
     * Get an order by its code
     * @param code The code of the order to retrieve
     */
    getByCode(code) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/orders/code/${code}`);
            return response.data;
        });
    }
    /**
     * Delete an order
     * @param orderId The ID of the order to delete
     * @param isVoidPayment Whether to void the associated payment
     */
    delete(orderId_1) {
        return __awaiter(this, arguments, void 0, function* (orderId, isVoidPayment = false) {
            yield this.client.apiClient.delete(`/orders/${orderId}`, {
                params: {
                    IsVoidPayment: isVoidPayment,
                },
            });
        });
    }
    /**
     * Get orders by date range
     * @param fromDate Start date (YYYY-MM-DD)
     * @param toDate End date (YYYY-MM-DD)
     * @param params Additional filter parameters
     */
    getByDateRange(fromDate_1, toDate_1) {
        return __awaiter(this, arguments, void 0, function* (fromDate, toDate, params = {}) {
            const response = yield this.client.apiClient.get('/orders', {
                params: Object.assign(Object.assign({}, params), { fromPurchaseDate: fromDate, toPurchaseDate: toDate }),
            });
            return response.data;
        });
    }
    /**
     * Get orders by customer
     * @param customerIdentifier Customer's phone number or code
     * @param params Additional filter parameters
     */
    getByCustomer(customerIdentifier_1) {
        return __awaiter(this, arguments, void 0, function* (customerIdentifier, params = {}) {
            // Try to determine if the identifier is a phone number
            const isPhone = /^\d+$/.test(customerIdentifier);
            const response = yield this.client.apiClient.get('/orders', {
                params: Object.assign(Object.assign({}, params), (isPhone ? { customerPhone: customerIdentifier } : { customerCode: customerIdentifier })),
            });
            return response.data;
        });
    }
}
exports.OrderHandler = OrderHandler;
