// kiotviet-client-sdk/dist/resources/products.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductHandler = void 0;
class ProductHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List products with optional filtering
     * @param params Filter parameters (pageSize, currentItem, code, name, categoryId, etc.)
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/products', { params });
            return response.data;
        });
    }
    /**
     * Get a product by its ID
     * @param productId The ID of the product to retrieve
     */
    getById(productId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/products/${productId}`);
            return response.data;
        });
    }
    /**
     * Create a new product
     * @param productData The product data to create
     */
    create(productData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/products', productData);
            return response.data;
        });
    }
    /**
     * Update an existing product
     * @param productId The ID of the product to update
     * @param productData The product data to update
     */
    update(productId, productData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/products/${productId}`, productData);
            return response.data;
        });
    }
    /**
     * Delete a product
     * @param productId The ID of the product to delete
     */
    delete(productId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.delete(`/products/${productId}`);
        });
    }
    /**
     * Get products by category ID
     * @param categoryId The ID of the category
     * @param params Additional filter parameters
     */
    getByCategory(categoryId_1) {
        return __awaiter(this, arguments, void 0, function* (categoryId, params = {}) {
            const response = yield this.client.apiClient.get('/products', {
                params: Object.assign(Object.assign({}, params), { categoryId }),
            });
            return response.data;
        });
    }
    /**
     * Search products by name or code
     * @param query Search query (name or code)
     * @param params Additional filter parameters
     */
    search(query_1) {
        return __awaiter(this, arguments, void 0, function* (query, params = {}) {
            const response = yield this.client.apiClient.get('/products', {
                params: Object.assign(Object.assign({}, params), { name: query }),
            });
            return response.data;
        });
    }
    /**
     * Get a product by its barcode
     * @param barcode The product barcode
     */
    /**
     * Get a product by its code
     * @param code The product code
     */
    getByCode(code) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/products/code/${code}`);
            return response.data;
        });
    }
    /**
     * Get all product attributes with their distinct values
     */
    getAttributes() {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get('/attributes/allwithdistinctvalue');
            return response.data;
        });
    }
    /**
     * Add multiple products at once
     * @param products Array of product data to create
     */
    bulkCreate(products) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.post('/listaddproducts', {
                listProducts: products,
            });
        });
    }
    /**
     * Update multiple products at once
     * @param products Array of product data to update
     */
    bulkUpdate(products) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.put('/listupdatedproducts', {
                listProducts: products,
            });
        });
    }
    /**
     * Get product inventory levels across branches
     * @param params Query parameters (branchIds, lastModifiedFrom, etc)
     */
    getInventoryLevels() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/productOnHands', { params });
            return response.data;
        });
    }
    getByBarcode(barcode) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get('/products', {
                params: {
                    barcode,
                    pageSize: 1,
                },
            });
            if (!response.data.data.length) {
                throw new Error(`Product with barcode ${barcode} not found`);
            }
            return response.data.data[0];
        });
    }
}
exports.ProductHandler = ProductHandler;
