// kiotviet-client-sdk/dist/resources/purchase-orders.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PurchaseOrderHandler = void 0;
const purchase_order_1 = require("../types/purchase-order");
class PurchaseOrderHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List purchase orders with optional filtering
     * @param params Filter parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/purchaseorders', { params });
            return response.data;
        });
    }
    /**
     * Get a purchase order by its ID
     * @param purchaseOrderId The ID of the purchase order to retrieve
     */
    getById(purchaseOrderId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/purchaseorders/${purchaseOrderId}`);
            return response.data;
        });
    }
    /**
     * Create a new purchase order
     * @param purchaseOrderData The purchase order data to create
     */
    create(purchaseOrderData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/purchaseorders', purchaseOrderData);
            return response.data;
        });
    }
    /**
     * Update an existing purchase order
     * @param purchaseOrderId The ID of the purchase order to update
     * @param purchaseOrderData The purchase order data to update
     */
    update(purchaseOrderId, purchaseOrderData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/purchaseorders/${purchaseOrderId}`, Object.assign({ id: purchaseOrderId }, purchaseOrderData));
            return response.data;
        });
    }
    /**
     * Cancel a purchase order
     * @param purchaseOrderId The ID of the purchase order to cancel
     * @param reason Optional cancellation reason
     */
    cancel(purchaseOrderId, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/purchaseorders/${purchaseOrderId}`, {
                id: purchaseOrderId,
                status: purchase_order_1.PurchaseOrderStatus.Cancelled,
                description: reason,
            });
            return response.data;
        });
    }
    /**
     * Get purchase orders by date range
     * @param fromDate Start date (YYYY-MM-DD)
     * @param toDate End date (YYYY-MM-DD)
     * @param params Additional filter parameters
     */
    getByDateRange(fromDate_1, toDate_1) {
        return __awaiter(this, arguments, void 0, function* (fromDate, toDate, params = {}) {
            const response = yield this.client.apiClient.get('/purchaseorders', {
                params: Object.assign(Object.assign({}, params), { fromPurchaseDate: fromDate, toPurchaseDate: toDate }),
            });
            return response.data;
        });
    }
    /**
     * Get purchase orders by supplier code
     * @param supplierCode The supplier code to filter by
     * @param params Additional filter parameters
     */
    getBySupplier(supplierCode_1) {
        return __awaiter(this, arguments, void 0, function* (supplierCode, params = {}) {
            const response = yield this.client.apiClient.get('/purchaseorders', {
                params: Object.assign(Object.assign({}, params), { supplierCode }),
            });
            return response.data;
        });
    }
}
exports.PurchaseOrderHandler = PurchaseOrderHandler;
