// kiotviet-client-sdk/dist/resources/surcharges.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SurchargeHandler = void 0;
class SurchargeHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List surcharges with optional filtering and pagination
     * @param params Filter and pagination parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/surcharges', { params });
            return response.data;
        });
    }
    /**
     * Get a surcharge by its ID
     * @param surchargeId The ID of the surcharge to retrieve
     */
    getById(surchargeId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/surcharges/${surchargeId}`);
            return response.data;
        });
    }
    /**
     * Create a new surcharge
     * @param surchargeData The surcharge data to create
     */
    create(surchargeData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/surcharges', surchargeData);
            return response.data;
        });
    }
    /**
     * Update an existing surcharge
     * @param surchargeId The ID of the surcharge to update
     * @param surchargeData The surcharge data to update
     */
    update(surchargeId, surchargeData) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/surcharges/${surchargeId}`, surchargeData);
            return response.data;
        });
    }
    /**
     * Delete a surcharge
     * @param surchargeId The ID of the surcharge to delete
     */
    delete(surchargeId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.delete(`/surcharges/${surchargeId}`);
        });
    }
}
exports.SurchargeHandler = SurchargeHandler;
