// kiotviet-client-sdk/dist/resources/transfers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferHandler = void 0;
class TransferHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * Get list of transfers with optional filtering and pagination
     * @param params Query parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/transfers', { params });
            return response.data;
        });
    }
    /**
     * Get transfer by ID
     * @param id Transfer ID
     */
    getById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/transfers/${id}`);
            return response.data;
        });
    }
    /**
     * Create a new transfer
     * @param data Transfer creation data
     */
    create(data) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.post('/transfers', data);
            return response.data;
        });
    }
    /**
     * Update transfer by ID
     * @param id Transfer ID
     * @param data Transfer update data
     */
    update(id, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.put(`/transfers/${id}`, data);
            return response.data;
        });
    }
    /**
     * Delete transfer by ID
     * @param id Transfer ID
     */
    delete(id) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.apiClient.delete(`/transfers/${id}`);
        });
    }
}
exports.TransferHandler = TransferHandler;
