// kiotviet-client-sdk/dist/resources/users.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserHandler = void 0;
class UserHandler {
    constructor(client) {
        this.client = client;
    }
    /**
     * List users with optional filtering
     * @param params Filter parameters
     */
    list() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/users', { params });
            return response.data;
        });
    }
    /**
     * Get a user by their ID
     * @param userId The ID of the user to retrieve
     */
    getById(userId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.client.apiClient.get(`/users/${userId}`);
            return response.data;
        });
    }
    /**
     * Get active users
     * @param params Additional filter parameters (excluding isActive)
     */
    getActive() {
        return __awaiter(this, arguments, void 0, function* (params = {}) {
            const response = yield this.client.apiClient.get('/users', {
                params: Object.assign(Object.assign({}, params), { isActive: true }),
            });
            return response.data;
        });
    }
    /**
     * Get users by branch
     * @param branchId The ID of the branch to filter by
     * @param params Additional filter parameters (excluding branchId)
     */
    getByBranch(branchId_1) {
        return __awaiter(this, arguments, void 0, function* (branchId, params = {}) {
            const response = yield this.client.apiClient.get('/users', {
                params: Object.assign(Object.assign({}, params), { branchId }),
            });
            return response.data;
        });
    }
    /**
     * Search users by name or code
     * @param query Search query (matches against name or code)
     * @param params Additional filter parameters (excluding name and code)
     */
    search(query_1) {
        return __awaiter(this, arguments, void 0, function* (query, params = {}) {
            const response = yield this.client.apiClient.get('/users', {
                params: Object.assign(Object.assign({}, params), { name: query }),
            });
            return response.data;
        });
    }
}
exports.UserHandler = UserHandler;
